src/platform/ESP32/DiagnosticDataProviderImpl.cpp

```diff
@@ -287,6 +287,61 @@ void DiagnosticDataProviderImpl::ReleaseNetworkInterfaces(NetworkInterface * net
     }
 }
 
+CHIP_ERROR DiagnosticDataProviderImpl::GetThreadMetrics(ThreadMetrics ** threadMetricsOut)
+{
+#ifdef CONFIG_FREERTOS_USE_TRACE_FACILITY
+    TaskStatus_t * taskStatusArray = nullptr;
+    ThreadMetrics * head           = nullptr;
+    uint32_t arraySize             = uxTaskGetNumberOfTasks();
+
+    taskStatusArray = (TaskStatus_t *) chip::Platform::MemoryAlloc(arraySize * sizeof(TaskStatus_t));
```

Comment:
Please consider using `MemoryCalloc` here.  Or better yet ScopedMemoryBuffer, to avoid the leaks you have right now.